File: Controlador/InventarioControlador.py
class InventarioControlador:
    def __init__(self, app):
        self.app = app
        self.panel = None

    def getApp(self):
        return self.app

    def setPanel(self, panel):
        self.panel = panel

    def filtrar(self, texto, categoria):
        todos = self.app.getProductos()
        resultado = []
        for p in todos:
            matchTexto = not texto or not texto.strip() or texto.lower() in p.getNombre().lower() or texto in str(p.getId())
            matchCat = not categoria or not categoria.strip() or p.getCategoria() == categoria
            if matchTexto and matchCat:
                resultado.append(p)
        return resultado

    def guardarProducto(self, p):
        existente = self.app.getProductoBD().buscarPorId(p.getId())
        if existente is not None:
            self.app.getProductoBD().actualizar(p)
        else:
            self.app.getProductoBD().insertar(p)
        if self.panel is not None:
            self.panel.refrescar()

    def eliminarProducto(self, id_prod):
        self.app.getProductoBD().eliminar(id_prod)
        if self.panel is not None:
            self.panel.refrescar()

    def buscarPorId(self, id_prod):
        return self.app.getProductoBD().buscarPorId(id_prod)

    def generarNuevoId(self):
        return self.app.getProductoBD().generarNuevoId()

File: Controlador/InventarioControlador.cs
using System;
using System.Collections.Generic;
using System.Linq;
using punto_de_venta_C_.Modelo;
using punto_de_venta_C_.Vista;

namespace punto_de_venta_C_.Controlador
{
    public class InventarioControlador
    {
        private readonly ControladorPrincipal app;
        private InventarioPanel panel;

        public InventarioControlador(ControladorPrincipal app)
        {
            this.app = app;
        }

        public ControladorPrincipal getApp()
        {
            return app;
        }

        public void setPanel(InventarioPanel panel)
        {
            this.panel = panel;
        }

        public List<Producto> filtrar(string texto, string categoria)
        {
            List<Producto> todos = app.getProductos();
            List<Producto> resultado = new List<Producto>();
            foreach (Producto p in todos)
            {
                bool matchTexto = string.IsNullOrEmpty(texto)
                        || p.getNombre().ToLower().Contains(texto.ToLower()) || p.getId().Contains(texto);
                bool matchCat = string.IsNullOrEmpty(categoria) || p.getCategoria().Equals(categoria);
                if (matchTexto && matchCat)
                    resultado.Add(p);
            }
            return resultado;
        }

        public void guardarProducto(Producto p)
        {
            Producto existente = app.getProductoBD().buscarPorId(p.getId());
            if (existente != null)
            {
                app.getProductoBD().actualizar(p);
            }
            else
            {
                app.getProductoBD().insertar(p);
            }
            if (panel != null)
                panel.refrescar();
        }

        public void eliminarProducto(string id)
        {
            app.getProductoBD().eliminar(id);
            if (panel != null)
                panel.refrescar();
        }

        public Producto buscarPorId(string id)
        {
            return app.getProductoBD().buscarPorId(id);
        }

        public string generarNuevoId()
        {
            return app.getProductoBD().generarNuevoId();
        }
    }
}
